[Control Center] Add identity providers section (#3907)
vaadin/docs: articles/control-center/identity-management/index.adoc

---

= Identity Management

Control Center provides powerful tools to manage users, groups, roles and identity providers for Vaadin applications deployed in a Kubernetes cluster. These features are essential for controlling access to various parts of an application and ensuring secure authentication and authorization for users.

With Control Center, you can do several things that allow you to centralize user management, enforce security policies, and control access across different parts of your Vaadin application:

- *Manage Users*: Create and manage users, enable or disable them, request password changes, or enable Multi-Factor Authentication (MFA).
- *Create & Assign Groups*: Organize users into groups, assign specific roles to the groups, and streamline permissions management.
- *Manage Roles*: Define roles that can be used to control access to specific views or functionality in your application.

> - *Manage Identity Providers*: Create identity providers, like social providers, to manage the authentication and authorization of the users.

Author: No, they are not mapped by default. I removed the authorization from the description.

In the [guilabel]*Identity Management* section, navigate to [guilabel]*Roles* and click the [guibutton]*Create* button to open the role creation form.

[.device]
image::images/role-create.png[Create Role Form]

Fill in the fields in the role creation form:

- *Name*: The name of the role (e.g., `ADMIN`).
- *Description*: A brief description of what this role entails (e.g., "Role for administrators.").

Click [guibutton]*Create* to save the new role. This role can now be assigned to users or to groups to control access.

== Creating a Group & Assigning a Role

> To manage efficiently the permissions, you can organize users into groups and assign roles to these groups. This saves you the time of having to assign the same privileges to each person. For instance, you might have all sales people in the group called, _sales_ and all managers in another group called, _managers_. And if you have a few users who are sales managers, you can assign them to both groups.

Reviewer: @russelljtdyer 
Perhaps this would be easier to read?
```suggestion
To manage the permissions efficiently, you can organize users into groups and assign roles to these groups. This saves you the time of having to assign the same privileges to each person. For instance, you might have all sales people in the group called, _sales_ and all managers in another group called, _managers_. And if you have a few users who are sales managers, you can assign them to both groups.
```

- *Roles*: Select the roles for the group (e.g., `ADMIN`).
- *Members*: Add users to the group by selecting them from the list of users.

Once the group is created, the assigned users inherit the permissions associated with the roles assigned to the group.

== Creating Identity Providers

An identity provider represents an external service, which can be used to authenticate access to Control Center, or a deployed application. This can be a social provider or a custom provider. The supported social providers are GitHub, Google, LinkedIn, Microsoft, and X (formerly Twitter). A custom provider can be any that implements the OpenID Connect 1.0 protocol (e.g., Okta).

When creating an identity provider, you need to choose the type of the provider you want to create from the list presented. After, in the case of social providers, you need to specify the `Alias`, `Display Name`, `Client ID`, and `Client Secret`. If you already have a client in your external identity provider, you only need to use its ID and secret. If not, you can follow the link under the provider list, which takes you to the provider's page where you can create a new client. 

> When everything is filled in, you can copy the `Login Redirect URI` or the `Logout Redirect URI` and use them in your external provider's client to tell it where to redirect the user after a successful authentication or logout. At the end, click on the [guibutton]*Create* button. As an example, creating a Google identity provider looks like this:

Reviewer: @russelljtdyer 
```suggestion
When everything is filled in, you can copy the `Login Redirect URI` or the `Logout Redirect URI` and use them in your external provider's client to tell it where to redirect the user after a successful authentication or logout, respectively. At the end, click on the [guibutton]*Create* button. As an example, creating a Google identity provider looks like this:
```

[.device]
image::images/provider-login.png[Identity Provider Login Form]

When you click on one of the providers, it redirects you to the external provider's login page where you need to specify your external user's username and password:

[.device]
image::images/okta-login.png[Custom Identity Provider Login]

After a successful authentication, if a user with the same email already exists in Control Center, you can link that user with one or more identities from identity providers, or create a new user based on the identity information from them:

[.device]
image::images/link-user-profile.png[Link User Profile Form]

If you click on the [guibutton]*Review Profile* button, you can edit the identity information before linking it to a user in Control Center:

[.device]
image::images/review-user-profile.png[Review User Profile Form]

After successfully linking a user and an external identity, you just need to confirm this linking by logging in with your user:

[.device]
image::images/confirm-user-profile.png[Confirm User Profile Form]

The authentication flow behind these steps is the following:

1. The unauthenticated user requests a protected view in an application.

> 2. The application redirects the user to Control Center to authenticate.

Author: I would say yes. Control Center hides Keycloak from the user, it even creates its own theme for login forms. If this approach is acceptable, then we can say that Control Center handles the authentication.

> 2. The application redirects the user to Control Center to authenticate.

Reviewer 2: > Control Center handles the authentication

This ☝️ No need to mention Keycloak until we provide documentation to access its admin console.

> 1. The unauthenticated user requests a protected view in an application.
> 2. The application redirects the user to Control Center to authenticate.
> 3. Control Center displays the login page with a list of configured identity providers.
> 4. The user selects one of the identity providers by clicking its button.
> 5. Control Center issues an authentication request to the target identity provider requesting authentication and redirects the user to the identity provider’s login page.
> 6. The user provides credentials or consents to authenticate with the identity provider.
> 7. Upon successful authentication by the identity provider, the user redirects back to Control Center with an authentication response.
> 8. Control Center checks if the response from the identity provider is valid. If valid, Control Center imports and creates a user if the user does not already exist. If the user already exists, Control Center asks the user to link the identity returned from the identity provider with the existing user. This behavior is account linking.
> 9. When the user authenticates, Control Center permits access to the protected view.

Reviewer: ```suggestion
. The unauthenticated user requests a protected view in an application.
. The application redirects the user to Control Center to authenticate.
. Control Center displays the login page with a list of configured identity providers.
. The user selects one of the identity providers by clicking its button.
. Control Center issues an authentication request to the target identity provider requesting authentication and redirects the user to the identity provider’s login page.
. The user provides credentials or consents to authenticate with the identity provider.
. Upon successful authentication by the identity provider, the user redirects back to Control Center with an authentication response.
. Control Center checks if the response from the identity provider is valid. If it is valid, one of two things happen:
.. If the user already exists, Control Center asks the user to link the accounts (i.e. link the identity returned from the identity provider with the existing user).
.. If the user does not exist, Control Center imports and creates a user.
. When the user authenticates, Control Center permits access to the protected view.
```

When creating an identity provider, you need to choose the type of the provider you want to create from the list. After, in the case of social providers, you need to specify the `Alias`, `Display Name`, `Client ID`, and `Client Secret`. If you already have a client in your external identity provider, you only need to use its ID and secret. If not, then you can follow the link under the provider list, which takes you to the provider's page, where you can create a new client. When everything is filled out, you can copy the `Login Redirect URI` and/or the `Logout Redirect URI` and use it in your external provider's client to let it know where to redirect the user after a successful authentication or logout. At the end, click on the [guibutton]*Create* button. As an example, creating a *Google* identity provider looks like this:

[.device]
image::images/provider-create.png[Identity Provider Create Form]

After creating an identity provider, you can find it in the *Identity Providers* list. You can select one to check its details, update some of the fields, or delete it:

[.device]
image::images/provider-detail.png[Identity Provider Detail Form]

> When you want to create a custom *OpenID Connect 1.0* provider, you also need to specify the `Discovery Endpoint`. This is used during the provider creation to get relevent metadata from the external provider. The endpoint URL has a specific format: it always ends with the `.well-known/openid-configuration` path. (e.g. https://1234-admin.okta.com/.well-known/openid-configuration). Usually, you can find the discovery endpoint in the client details after you create it on the external provider. As an example, creating an *Okta* provider looks like this:

Reviewer: ```suggestion
When you want to create a custom *OpenID Connect 1.0* provider, you also need to specify the `Discovery Endpoint`. This is used during the provider creation to get relevant metadata from the external provider. The endpoint URL has a specific format: it always ends with the `.well-known/openid-configuration` path. (e.g. https://1234-admin.okta.com/.well-known/openid-configuration). Usually, you can find the discovery endpoint in the client details after you create it on the external provider. As an example, creating an *Okta* provider looks like this:
```

[.device]
image::images/review-user-profile.png[Review User Profile Form]

After successfully linking a user and an external identity, you need to confirm this link by logging in with your user:

[.device]
image::images/confirm-user-profile.png[Confirm User Profile Form]

The authentication process behind these steps is the following:

- The unauthenticated user requests a protected view in an application.
- The application redirects the user to Control Center to authenticate, where it displays the login page with a list of configured identity providers.
- The user selects one of the identity providers by clicking its respective button.
- Control Center issues an authentication request to the target identity provider, asking for authentication and redirects the user to the identity provider’s login page.
- The user provides credentials or consents to authenticate with the identity provider.
- Upon successful authentication by the identity provider, the user is redirected back to Control Center with an authentication response.

> - Control Center checks if the response from the identity provider is valid. If it is, one of two things happen: If the user already exists, Control Center asks the user to link the accounts (i.e., link the identity returned from the identity provider with the existing user). If the user doesn't exist, Control Center imports and creates a user.

Reviewer: @russelljtdyer 
What's the reasoning behind removing the numbering? This is a sequence of events, not a list of related items.

Reviewer: Thank you for the explanation.
